From 40b3796325c290db1bd7ac90b8dcbfa0caf790e6 Mon Sep 17 00:00:00 2001
From: JayZhang-wiwynn <jay_zhang@wiwynn.com>
Date: Thu, 11 Jul 2024 18:11:00 +0800
Subject: [PATCH] greatlakes:olympic 2.0: Add OP2 Platinfo for support auto
 crashdump v0.9

Description:
 Add OP2 platform information for support intel ACD v0.9

Motivation:
 Let BMC support intel auto crashdump v0.9
---
 PlatInfo.cpp | 110 +++++++++++++++++++++++++++++++++++++++++++++++----
 1 file changed, 102 insertions(+), 8 deletions(-)

diff --git a/PlatInfo.cpp b/PlatInfo.cpp
index 410695a..58bdfc5 100644
--- a/PlatInfo.cpp
+++ b/PlatInfo.cpp
@@ -19,27 +19,121 @@
 
 extern "C" {
 #include <stdio.h>
+#include "engine/utils.h"
 }
 #include <string>
+#include <sstream>
+#include <vector>
+
+using namespace std;
+
+#define MAX_BUF_SIZE 128
 
-#ifdef IPMB_PECI_INTF
 extern int node_bus_id;
-#endif
+extern string crashdumpFruName;
+
+static void split(const string &s, vector<string> &sv, const char delim = ' ') {
+  istringstream iss(s);
+  string temp;
+
+  sv.clear();
+  while (getline(iss, temp, delim)) {
+    sv.emplace_back(move(&temp[2]));
+  }
+}
+
+static int get_fw_ver(const string &fru, const string &comp, string &ver) {
+  char line[MAX_BUF_SIZE];
+  FILE *fp;
+  string cmd;
+
+  cmd = "/usr/bin/fw-util " + fru + " --version " + comp;
+  if (!(fp = popen(cmd.c_str(), "r")))
+    return -1;
+
+  if (fgets(line, sizeof(line), fp)) {
+    pclose(fp);
+
+    string str = line;
+    auto start = str.find(":");
+    if (start == string::npos)
+      return -1;
+
+    start = str.find_first_not_of(" ", start+1);
+    if (start == string::npos)
+      return -1;
+
+    auto end = str.find("\n", start);
+    if (end == string::npos) {
+      end = str.size();
+    }
 
-extern std::string crashdumpFruName;
+    ver = str.substr(start, end - start);
+  }
+
+  return 0;
+}
 
 void getBmcVersion(char *ver, size_t) {
-  (void)ver;
+  string ver_str;
+
+  if (get_fw_ver("bmc", "bmc", ver_str) == 0) {
+    cd_snprintf_s(ver, SI_BMC_VER_LEN, "%s", ver_str.c_str());
+  }
 }
 
 void getBiosVersion(char *ver, size_t) {
-  (void)ver;
+  string ver_str;
+  string slot_str = "slot" + to_string(node_bus_id+1);
+
+  if (get_fw_ver(slot_str, "bios", ver_str) == 0) {
+    cd_snprintf_s(ver, SI_BIOS_ID_LEN, "%s", ver_str.c_str());
+  }
 }
 
-void getSystemGuid(std::string &guid) {
-  guid = "";
+void getSystemGuid(string &guid) {
+  char line[MAX_BUF_SIZE], cmd[MAX_BUF_SIZE];
+  FILE *fp;
+  string slot_str = "slot" + to_string(node_bus_id+1);
+  string guid_str("");
+
+  snprintf(cmd, sizeof(cmd), "/usr/local/bin/guid-util %s sys --get 2>/dev/null", slot_str.c_str());
+  if (!(fp = popen(cmd, "r")))
+    return;
+
+  if (fgets(line, sizeof(line), fp)) {
+    pclose(fp);
+
+    string str = line;
+    auto start = str[0];
+    if (start == string::npos)
+      return;
+
+    size_t pos = 0;
+    while ((pos = str.find("\n", pos)) != string::npos) {
+      str.erase (pos, 1);
+    }
+
+    vector<string> sv;
+    split(str, sv, ':');
+
+    // Hanle LSB
+    swap(sv[0], sv[3]);
+    swap(sv[1], sv[2]);
+    swap(sv[4], sv[5]);
+    swap(sv[6], sv[7]);
+    for (int i = 0; i < (int)sv.size(); i++) {
+      if (i == 4 || i == 6 || i == 8 || i == 10) {
+        guid_str += "-";
+      }
+      guid_str += sv[i];
+    }
+
+    guid = guid_str;
+  }
 }
 
 void platformInit(uint8_t fru) {
-  (void)fru;
+  node_bus_id = fru - 1;
+  crashdumpFruName = "slot" + to_string(fru) + "_";
 }
-- 
2.25.1

